### PR TITLE
#65 implementing investor startup direct contact functionality

### PR DESCRIPTION
**Pull request by olexiyko**
In this PR, I include tasks 39, 40, 36, and 42.I would also appreciate your comments on the consumer, as I’m not entirely sure if my implementation is good.

### CODE REVIEW

File: forum/communications/tests.py [hunk 1]
```diff
@@ -1,3 +1,199 @@
...
+
+        self.room.name = f"room_{self.user1.user_id}_{self.user2.user_id}"
+        self.room.save()
+        self.room.online.set([self.user1, self.user2])
+
+        expected_names = {
+            "user_1": self.user1.first_name,
+            "user_2": self.user2.first_name,
+        }
+        self.assertEqual(self.room.get_users_names(), expected_names)
```
Comment (mehalyna) [inline forum/communications/tests.py:50]: Add tests for edge cases, e.g., calling 'leave()' for a user not in the room.
Test 'get_online_count()' when no users are in the room.

_Examples_
```
def test_leave_nonexistent_user(self):
    """Ensure no error is raised when a user leaves a room they are not in."""
    self.room.leave(self.user1)
    self.assertNotIn(self.user1, self.room.online.all())

def test_get_online_count_empty_room(self):
    """Ensure online count is zero for an empty room."""
    self.assertEqual(self.room.get_online_count(), 0)
```

File: forum/communications/tests.py
Comment (mehalyna) [inline forum/communications/tests.py:1]: Consider adding tests for validating encrypted message content using the `Message` model.
Validate behavior for invalid or overly long message content.
_Examples:_
```
def test_message_encryption(self):
    """Test that message content is encrypted and decrypted correctly."""
    room = Room.objects.create(name="room_1_2")
    message = Message.objects.create(user=self.user1, room=room, content="Hello!")
    self.assertNotEqual(message.content, "Hello!")  # Content should be encrypted
    decrypted_message = Message.objects.get(pk=message.pk).content
    self.assertEqual(decrypted_message, "Hello!")  # Content should decrypt correctly

def test_message_max_length(self):
    """Test that overly long message content raises an error."""
    room = Room.objects.create(name="room_1_2")
    long_message = "x" * 513  # Exceeding max length
    with self.assertRaises(ValueError):
        Message.objects.create(user=self.user1, room=room, content=long_message)
```

File: forum/communications/tests.py [hunk 2]
```diff
@@ -1,3 +1,199 @@
...
+            f"Message '{message}' was not received",
+        )
+
+    @property
+    def _chat_log_value(self):
+        chat_log = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.ID, "chat-log")),
+            "Chat log element not found",
+        )
+        return chat_log.get_property("value")
```
Comment (mehalyna) [inline forum/communications/tests.py:165]: - Add tests for WebSocket disconnection and reconnection scenarios.
 - Test sending invalid JSON payloads over WebSocket.
 - Validate the handling of edge cases, such as attempting to join a non-existent room.
 
_Examples:_
```
def test_websocket_reconnection(self):
    """Test that WebSocket can reconnect after an unexpected closure."""
    self._enter_chat_room("room_1")
    self._switch_to_window(0)
    self.driver.execute_script("window.location.reload()")
    WebDriverWait(self.driver, 10).until(
        lambda driver: "room_1" in driver.current_url,
        "Failed to reconnect to the WebSocket room after reload",
    )

def test_send_invalid_json(self):
    """Test that sending invalid JSON data raises an error."""
    self._enter_chat_room("room_1")
    self.driver.execute_script("""
        const socket = new WebSocket('ws://' + window.location.host + '/ws/api/v1/communication/room_1/');
        socket.onopen = () => socket.send('Invalid JSON');
    """)
    error_message = WebDriverWait(self.driver, 10).until(
        lambda driver: "Unexpected error" in driver.page_source,
        "Error message not displayed for invalid JSON",
    )
    self.assertTrue(error_message)
```

File: forum/communications/tests.py [hunk 3]
```diff
@@ -1,3 +1,199 @@
...
+    def test_message_history(self):
+        room = Room.objects.create()
+        room.online.set([self.user1, self.user2])
+        Message.objects.create(room=room, user=self.user1, content="Hello!")
+        self.client.force_authenticate(user=self.user2)
+        response = self.client.get(
+            f"/api/v1/communications/conversations/{room.id}/messages/"
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
```
Comment (mehalyna) [inline forum/communications/tests.py:199]: - Test creating a room with invalid participant data.
 - Test deleting a room that doesn’t exist.
 - Validate proper HTTP responses for unauthorized access.

_Examples:_

```
def test_create_room_with_invalid_participants(self):
    """Test creating a room with invalid participant data."""
    self.client.force_authenticate(user=self.user1)
    response = self.client.post("/api/v1/communications/conversations/", {"online": [99999, 88888]})
    self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
    self.assertIn("participants", response.data)

def test_delete_non_existent_room(self):
    """Test deleting a non-existent room."""
    self.client.force_authenticate(user=self.user1)
    response = self.client.delete("/api/v1/communications/conversations/999/")
    self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```

File: forum/communications/tests.py
Comment (mehalyna) [inline forum/communications/tests.py:1]: Consider adding setup utilities:

 - Create a reusable factory for setting up test data.

```
def setup_users_and_rooms(self):
    self.user1 = User.objects.create_user(email="user1@example.com", password="password")
    self.user2 = User.objects.create_user(email="user2@example.com", password="password")
    self.room = Room.objects.create(name="room_1_2")
```

- Mock WebSocket connections to test WebSocket failures.

```
def mock_websocket_failure(self):
    self.driver.execute_script("""
        const socket = new WebSocket('ws://' + window.location.host + '/ws/api/v1/communication/nonexistent_room/');
        socket.onclose = () => console.log('WebSocket closed');
    """)
```

File: forum/communications/views.py
Comment (mehalyna) [inline forum/communications/views.py:1]: Currently, any authenticated user can access the `CreateConversationView`, `SendMessageView`, and `MessageHistoryView`. This might lead to unauthorized access to sensitive data.
Solution: Add custom permission classes to ensure only authorized participants of a conversation can perform actions.

_Example:_
```
from rest_framework.permissions import BasePermission

class IsParticipant(BasePermission):
    def has_permission(self, request, view):
        room_id = view.kwargs.get("conversation_id")
        room = Room.objects.filter(pk=room_id).first()
        return room and request.user in room.online.all()
```

Use this permission class for `MessageHistoryView`:

```
permission_classes = [IsAuthenticated, IsParticipant]
```

File: forum/communications/views.py [hunk 4]
```diff
@@ -1,3 +1,41 @@
...
+        serializer.save(sender=self.request.user)
+
+
+class MessageHistoryView(generics.ListAPIView):
+    serializer_class = MessageSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        room_id = self.kwargs["conversation_id"]
+        return Message.objects.filter(room_id=room_id).order_by("timestamp")
```
Comment (mehalyna) [inline forum/communications/views.py:41]: `Message.objects.filter(room_id=room_id)` may fetch unnecessary fields.
Use `only()` or `values()` to fetch only necessary fields for efficiency.

File: forum/communications/views.py [hunk 5]
```diff
@@ -1,3 +1,41 @@
...
+    def perform_create(self, serializer):
+        serializer.save()
+
+
+class SendMessageView(generics.CreateAPIView):
+    serializer_class = MessageSerializer
+    permission_classes = [IsAuthenticated]
+
+    def perform_create(self, serializer):
+        serializer.save(sender=self.request.user)
```
Comment (mehalyna) [inline forum/communications/views.py:32]: A user can potentially send a message to a room they are not part of.
Validate the room's participants before saving the message.

_Example:_

```
def perform_create(self, serializer):
    room = serializer.validated_data["room"]
    if self.request.user not in room.online.all():
        raise PermissionDenied("You are not a participant of this conversation.")
    serializer.save(user=self.request.user)
```

File: forum/communications/views.py (hunk 4 above)
Comment (mehalyna) [inline forum/communications/views.py:41]: If a room has a long message history, fetching all messages at once could be inefficient.
Implement pagination to handle large datasets.

File: forum/communications/views.py [hunk 6]
```diff
@@ -1,3 +1,41 @@
...
+def room(request, room_name):
+    return render(request, "communications/room.html", {"room_name": room_name})
+
+
+class CreateConversationView(generics.CreateAPIView):
+    serializer_class = ChatRoomSerializer
+    permission_classes = [IsAuthenticated]
+
+    def perform_create(self, serializer):
+        serializer.save()
```
Comment (mehalyna) [inline forum/communications/views.py:24]: Currently, no checks ensure the participants are valid users or not already in a conversation.
Add validation for unique conversations and valid participants.

_Example:_

```
from rest_framework.exceptions import ValidationError

def perform_create(self, serializer):
    participants = serializer.validated_data.get("participants")
    if len(participants) != 2:
        raise ValidationError("A conversation must have exactly two participants.")
    existing_room = Room.objects.filter(name=f"room_{participants[0].id}_{participants[1].id}").exists()
    if existing_room:
        raise ValidationError("This conversation already exists.")
    serializer.save()
```

File: forum/communications/views.py [hunk 7]
```diff
@@ -1,3 +1,41 @@
...
-# Create your views here.
+User = get_user_model()
+
+
+def index(request):
+    return render(request, "communications/index.html")
+
+
+def room(request, room_name):
+    return render(request, "communications/room.html", {"room_name": room_name})
```
Comment (mehalyna) [inline forum/communications/views.py:16]: Consider validation of `room_name` and handle cases where the room does not exist

File: forum/forum/asgi.py
Comment (mehalyna) [inline forum/forum/asgi.py:1]: Add logging for WebSocket errors to help debug issues.

_Example_

```
import logging

logger = logging.getLogger("django")

application = ProtocolTypeRouter(
    {
        "http": django_asgi_app,
        "websocket": AllowedHostsOriginValidator(
            AuthMiddlewareStack(
                URLRouter(websocket_urlpatterns)
            )
        ),
    }
)

# Log connection errors
class LoggingMiddleware:
    async def __call__(self, scope, receive, send):
        try:
            await super().__call__(scope, receive, send)
        except Exception as e:
            logger.error(f"WebSocket error: {e}")
            raise
```

File: forum/forum/asgi.py [hunk 8]
```diff
@@ -9,8 +9,23 @@
...
 from django.core.asgi import get_asgi_application
 
-os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'forum.settings')
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "forum.settings")
+# Initialize Django ASGI application early to ensure the AppRegistry
+# is populated before importing code that may import ORM models.
+django_asgi_app = get_asgi_application()
 
-application = get_asgi_application()
+from communications.routing import websocket_urlpatterns
```
Comment (mehalyna) [inline forum/forum/asgi.py:22]: Group imports logically:
 - Standard library imports
 - Third-party library imports
 - Project-specific imports

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/communications/consumers.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+class ChatConsumer(AsyncWebsocketConsumer):
+    async def connect(self):
+        self.room_name = self.scope["url_route"]["kwargs"]["room_name"]
+        self.room_group_name = f"chat_{self.room_name}"
+
+        # Join room group
+        await self.channel_layer.group_add(self.room_group_name, self.channel_name)
+
+        await self.accept()
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/consumers.py]: Done

File: forum/communications/models.py [hunk 10]
```diff
@@ -1,3 +1,49 @@
...
+    def get_users_names(self):
+        User = get_user_model()
+        return {
+            "user_1": User.objects.get(
+                user_id=self.get_users_id()["user_1"]
+            ).first_name,
+            "user_2": User.objects.get(
+                user_id=self.get_users_id()["user_2"]
+            ).first_name,
+        }
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/models.py]: Done

File: forum/communications/serializers.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    class Meta:
+        model = Room
+        fields = ["id", "participants"]
+
+
+class MessageSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Message
+        fields = ["id", "room", "user", "content", "timestamp"]
+        read_only_fields = ["user", "timestamp"]
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/serializers.py:30]: There is no need to impelement decryption because encrypted filed can already be decrypted by his own when we turn to him

File: forum/communications/templates/communications/room.html [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      const roomName = JSON.parse(
+        document.getElementById("room-name").textContent
+      );
+      const chatSocket = new WebSocket(
+        "ws://" +
+          window.location.host +
+          "/ws/api/v1/communication/" +
+          roomName +
+          "/"
+      );
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/templates/communications/room.html]: Done

File: forum/communications/templates/communications/room.html [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+      chatSocket.onmessage = function (e) {
+        const data = JSON.parse(e.data);
+        document.querySelector("#chat-log").value += data.message + "\n";
+      };
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/templates/communications/room.html]: Done

File: forum/communications/templates/communications/room.html [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+      document.querySelector("#chat-message-submit").onclick = function (e) {
+        const messageInputDom = document.querySelector("#chat-message-input");
+        const message = messageInputDom.value;
+        chatSocket.send(
+          JSON.stringify({
+            message: message,
+          })
+        );
```
[the remark this answers is not in the document]
Comment (olexiyko) [reply forum/communications/templates/communications/room.html]: Done

File: forum/communications/tests.py (hunk 1 above)
Comment (olexiyko) [reply forum/communications/tests.py:62]: Done

File: forum/communications/tests.py
Comment (olexiyko) [reply forum/communications/tests.py:1]: Done

File: forum/communications/tests.py (hunk 2 above)
Comment (olexiyko) [reply forum/communications/tests.py]: Done

File: forum/communications/tests.py (hunk 3 above)
Comment (olexiyko) [reply forum/communications/tests.py]: Done

File: forum/communications/views.py
Comment (olexiyko) [reply forum/communications/views.py:1]: Done

File: forum/communications/views.py (hunk 4 above)
Comment (olexiyko) [reply forum/communications/views.py]: Done

Comment (olexiyko) [reply forum/communications/views.py]: Done

File: forum/communications/views.py (hunk 6 above)
Comment (olexiyko) [reply forum/communications/views.py]: Done

File: forum/forum/asgi.py
Comment (olexiyko) [reply forum/forum/asgi.py:1]: Done

File: forum/forum/asgi.py (hunk 8 above)
Comment (olexiyko) [reply forum/forum/asgi.py]: Done

File: forum/communications/tests.py [hunk 15]
```diff
@@ -1,3 +1,273 @@
...
+
+        response = self.client.get(
+            f"/api/v1/communications/conversations/{self.room.id}/messages/",
+            HTTP_AUTHORIZATION=f"Bearer {self.token}",
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertIn("results", response.data)
+        self.assertEqual(len(response.data["results"]), 10)
+        self.assertIn("next", response.data)
+        self.assertIsNotNone(response.data["next"])
```
Comment (mehalyna) [inline forum/communications/tests.py:238]: Additional edge cases and assertions can make tests more robust, e.g.:
 - Validate behavior when there are fewer than 10 messages.
 - Validate the behavior for the last page (no `next` link).
 - Validate invalid page numbers (proper handling of non-existent pages)

File: forum/communications/tests.py [hunk 16]
```diff
@@ -1,3 +1,273 @@
...
+        response = self.client.post(
+            f"/api/v1/communications/messages/",
+            data=data,
+            HTTP_AUTHORIZATION=f"Bearer {self.token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        message = Message.objects.last()
+        self.assertEqual(message.content, data["content"])
+        self.assertEqual(message.room, self.room)
+        self.assertEqual(message.user, self.user1)
```
Comment (mehalyna) [inline forum/communications/tests.py:273]: For tests involving JWT or external services, ensure to mock unnecessary calls to improve test performance and avoid dependencies.
_Example_

```
from unittest.mock import patch

@patch("rest_framework_simplejwt.tokens.RefreshToken.for_user")
def test_jwt_mocked(self, mock_refresh_token):
    mock_refresh_token.return_value = "mocked_token"
    response = self.client.post("/auth/jwt/create/", {"email": self.user1.email, "password": "password"})
    self.assertEqual(response.status_code, 200)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/communications/tests.py (hunk 15 above)
Comment (olexiyko) [reply forum/communications/tests.py]: Done

File: forum/communications/tests.py (hunk 16 above)
Comment (olexiyko) [reply forum/communications/tests.py]: Done

**Review by mehalyna — APPROVED** [review]

**Review by SergeKry — APPROVED** [review]

**Review by djmelyk-debelyk — APPROVED** [review]
